File: sol/solution/0001-0100/0028. Find the Index of the First Occurrence in a String/Solution.php
class Solution {
    /**
     * @param String $haystack
     * @param String $needle
     * @return Integer
     */
    function strStr($haystack, $needle) {
        $n = strlen($haystack);
        $m = strlen($needle);
        if ($m == 0) return 0;
        $lps = array_fill(0, $m, 0);
        for ($i = 1, $len = 0; $i < $m; ) {
            if ($needle[$i] == $needle[$len]) {
                $lps[$i++] = ++$len;
            } elseif ($len > 0) {
                $len = $lps[$len - 1];
            } else {
                $lps[$i++] = 0;
            }
        }
        for ($i = 0, $j = 0; $i < $n; ) {
            if ($haystack[$i] == $needle[$j]) {
                $i++; $j++;
                if ($j == $m) return $i - $m;
            } elseif ($j > 0) {
                $j = $lps[$j - 1];
            } else {
                $i++;
            }
        }
        return -1;
    }
}

File: sol/solution/0001-0100/0028. Find the Index of the First Occurrence in a String/Solution.rb
# @param {String} haystack
# @param {String} needle
# @return {Integer}
def str_str(haystack, needle)
  n, m = haystack.length, needle.length
  return 0 if m == 0
  lps = Array.new(m, 0)
  len = 0; i = 1
  while i < m
    if needle[i] == needle[len]
      len += 1
      lps[i] = len
      i += 1
    elsif len > 0
      len = lps[len - 1]
    else
      lps[i] = 0
      i += 1
    end
  end
  i = j = 0
  while i < n
    if haystack[i] == needle[j]
      i += 1; j += 1
      return i - m if j == m
    elsif j > 0
      j = lps[j - 1]
    else
      i += 1
    end
  end
  -1
end